### PR TITLE
Whylog factory

### PR DESCRIPTION
**Pull request by andrzejgorski**

### CODE REVIEW

File: requirements-test.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+mock
```
Comment (epawlowska) [inline requirements-test.txt]: Please, keep this file sorted.

File: whylog/front/utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, path, host):
+        self.path = path
+        self.host = host
+
+    def __repr__(self):
+        return '%s:%s' % self.host, self.path
+
+
+class LocallyAccessibleLogOpener(AbstractResourceLocation):
+    """
```
Comment (epawlowska) [inline whylog/front/utils.py]: Does LocallyAccesibleLogOpener open logs? I can't deduce it from docstrings and implementation. I thought it only aggregates information about file. Maybe comment it or change a name?

File: whylog/front/whylog_factory.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+from whylog.assistant import RegexAssistant
+from whylog.config import YamlConfig
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+
+
+def _teachers_generator(config, pattern_assistant):
+    """
+    Every call of this function return new teacher for new lesson.
```
Comment (epawlowska) [inline whylog/front/whylog_factory.py]: typo: return -> returns

File: whylog/front/whylog_factory.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    """
+    while True:
+        yield Teacher({0: ''}, 0, config, pattern_assistant)
+
+
+def whylog_factory(path, pattern_assistant=None):
+    """
+    This function create object of YamlConfig and send this object
```
Comment (epawlowska) [inline whylog/front/whylog_factory.py]: type: create -> creates, send -> sends

Comment (epawlowska) [inline whylog/front/whylog_factory.py]: "send" is confusing to me. This function doesn't send YamlConfig anywhere. It uses YamlConfig to create LogReader and teacher_generator.

File: whylog/front/whylog_factory.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    to the LogReader object and teachers_generator.
+    """
+    if pattern_assistant is None:
+        pattern_assistant = RegexAssistant()
+
+    with open(path, 'r') as raw_config:
```
Comment (epawlowska) [inline whylog/front/whylog_factory.py]: Please, rename `path` so that it tells more.

File: whylog/tests/tests_config/__init__.py [hunk 6]
```diff
@@ -7,6 +7,7 @@
 from whylog.config.parsers import RegexParserFactory
 from whylog.config.rule import RegexRuleFactory
 from whylog.teacher.user_intent import UserConstraintIntent, UserParserIntent, UserRuleIntent
+from whylog.tests.utils import path_test_files
 
 # Constraint types
 identical_constr = "identical"
```
Comment (epawlowska) [inline whylog/tests/tests_config/__init__.py:19]: Maybe put somewhere const constraint types. Maybe in whylog/constraints/consts.py? 

There are now 3 files where constraint types are defined: this file, whylog/teacher/mock_outputs.py, whylog/tests/test_config/**init**.py. Maybe it would be worth to import these constraint types from one file?

File: whylog/tests/tests_front/tests_utils.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+def tests_unit_creation_sample_front_input():
+    offset = 123
+    line_content = 'I like trains'
+    path = '/sample/path/to/file'
+    resource_location = LocallyAccessibleLogOpener(path)
+    front_input = FrontInput(offset, line_content, resource_location)
+
+    assert front_input.offset == offset
+    assert front_input.line_content == line_content
+    assert front_input.resource_location is resource_location
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_utils.py]: Maybe remove this test? It tests only constructor. Bugs in constructor will quickly come out in any serious test which uses FrontInput.

File: whylog/tests/tests_front/tests_utils.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def tests_unit_creation_resource_location():
+    path = '/sample/path/to/file'
+    resource_location = LocallyAccessibleLogOpener(path)
+    assert ('%s' % resource_location) == path
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_utils.py]: Maybe remove this test? It seems to be very, very simple. It tests only constructor and `__repr__` function. I suppose it is not very valuable.

File: whylog/tests/utils.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+def create_yaml_paths():
+    path = os.path.join(*path_test_files)
+    config_path = os.path.join(path, 'config.yaml')
+    parsers_path = os.path.join(path, 'parsers.yaml')
+    rules_path = os.path.join(path, 'rules.yaml')
+    log_type_path = os.path.join(path, 'log_type.yaml')
+    return config_path, parsers_path, rules_path, log_type_path
+
+
+def create_config():
```
Comment (epawlowska) [inline whylog/tests/utils.py]: create_config creates yaml config. Please, rename to `create_yaml_config`, or make it parametrized

Comment (epawlowska) [inline whylog/tests/utils.py]: It returns 2 objects: config and path. Looking at the name `create_config`, I would expect only config. Please, fix it somehow. Maybe it is too much for one function?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    file_mock.return_value = yaml_config
+    with patch('yaml.load', file_mock, create=True):
+        log_reader, teacher_generator = whylog_factory(config_path)
+    teacher = next(teacher_generator)
+
+    # check if returned object classes
+    assert isinstance(log_reader, LogReader)
+    assert isinstance(teacher, Teacher)
+
+    # check other teachers
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Why do we test other teachers? Isn't it enough to check it once? Maybe remove 4 following lines?

File: whylog/front/utils.py (hunk 2 above)
Comment (andrzejgorski) [inline whylog/front/utils.py]: I'm not sure if it really will open the logs. The @kgromadzki told me to write this class and it will be used in Config. I should ask him what this class really will do.

File: whylog/tests/tests_config/__init__.py (hunk 6 above)
Comment (andrzejgorski) [inline whylog/tests/tests_config/__init__.py:19]: This is not what my PR do. I only move path_test_files to utils.py file becouse I need it in different place. Ask @kgromadzki to change it.

File: whylog/tests/tests_front/tests_utils.py (hunk 7 above)
Comment (andrzejgorski) [inline whylog/tests/tests_front/tests_utils.py]: If we haven't other tests I think should can writes unit tests. I will keep it.

File: whylog/tests/utils.py (hunk 9 above)
Comment (andrzejgorski) [inline whylog/tests/utils.py]: Create config function creates mock content of config file and path to this file. In my opinion there are nothing dirty with creating two objects. Every time when SB will need config file, he will need path to the file too. This is only simple function for tests. It isn't so much important.

File: whylog/tests/tests_front/tests_utils.py (hunk 8 above)
Comment (andrzejgorski) [inline whylog/tests/tests_front/tests_utils.py]: I don't want to delete unittests.

File: whylog/tests/tests_front/tests_whylog_factory.py (hunk 10 above)
Comment (andrzejgorski) [inline whylog/tests/tests_front/tests_whylog_factory.py]: I don't want to delete it.

File: whylog/front/utils.py (hunk 2 above)
Comment (kgromadzki) [inline whylog/front/utils.py]: This class represents location where we should find this file. In config/investigation_plan.py I created class LineSource. I recommend you to move its to Front and add import in config. Talk about it with @ppolewicz

File: whylog/tests/tests_front/tests_utils.py (hunk 8 above)
Comment (kgromadzki) [inline whylog/tests/tests_front/tests_utils.py]: It tests nothing. Also one above.

File: whylog/tests/tests_front/tests_whylog_factory.py (hunk 10 above)
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Please, give a reason why so?

File: whylog/front/utils.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        :param offset: byte of the first character of line in the file.
+        :param line_content: the content of the line.
+        :param resource_location: obeject of ResourceLocation.
+        """
+        self.offset = offset
+        self.line_content = line_content
+        self.resource_location = resource_location
+
+
+class AbstractResourceLocation(object):
```
Comment (ppolewicz) [inline whylog/front/utils.py]: This is not really an abstract class. You need to decorate it.

Also, this class does not belong in `whylog/front/utils.py`

File: whylog/front/utils.py (hunk 2 above)
Comment (ppolewicz) [inline whylog/front/utils.py]: Merge `LineSource` and `ResourceLocation`. Conceptually it is the same thing.

File: whylog/front/utils.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    The Abastract class of for agregating informations about
```
Comment (ppolewicz) [inline whylog/front/utils.py]: Typo in "Abstract"

File: whylog/front/whylog_factory.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+from whylog.assistant.const import AssistantType
+from whylog.assistant.regex_assistant import RegexAssistant
+from whylog.config import YamlConfig
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+
+
+def _teachers_generator(config, pattern_assistant):
+    """
+    Every call of this function returns new teacher for new lesson.
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: Pass a factory function instead of a generator.

File: whylog/front/whylog_factory.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+def create_pattern_assistant(config):
+    """
+    This function read from which pattern assistant should be created.
+    The default one now is RegexAssistant.
+    """
+    try:
+        if config['pattern_assistant'] == AssistantType.REGEX:
+            return RegexAssistant()
+    except KeyError:
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: This function is wrong. Please go back to first phase of review and focus on what it does and on what it was supposed to do.

File: whylog/front/whylog_factory.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        yield Teacher(config, pattern_assistant)
+
+
+def create_pattern_assistant(config):
+    """
+    This function read from which pattern assistant should be created.
+    The default one now is RegexAssistant.
+    """
+    try:
+        if config['pattern_assistant'] == AssistantType.REGEX:
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: Whylog front module shall not touch the yaml structure. Move the responsibility to config layer

File: whylog/front/whylog_factory.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def whylog_factory(config_file_path):
+    """
+    This function creates object of YamlConfig and use this object
+    to creates the LogReader object and teachers_generator.
+    """
+
+    with open(config_file_path, 'r') as raw_config:
+        config = yaml.load(raw_config)
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: What if the config is stored in a database?

Move the responsibility to config layer

File: whylog/front/whylog_factory.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+        pattern_assistant = create_pattern_assistant(config)
+        config_config = YamlConfig(
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: Please rename that variable

File: whylog/tests/tests_front/tests_utils.py (hunk 7 above)
Comment (ppolewicz) [inline whylog/tests/tests_front/tests_utils.py]: Remove it. It introduces fragility, also sometimes called "unit test hell"

File: whylog/tests/tests_front/tests_utils.py (hunk 8 above)
Comment (ppolewicz) [inline whylog/tests/tests_front/tests_utils.py]: Remove it please. Only test things which are worth testing.

File: whylog/tests/utils.py (hunk 9 above)
Comment (ppolewicz) [inline whylog/tests/utils.py]: Your layer should never touch config file path. If it did, current function name would be inappropriate.

File: whylog/front/utils.py [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+    """
+    Object of this class represents line in the file.
+    """
+
+    def __init__(self, offset, line_content, line_source):
+        """
+        Basic constructor of FrontInput.
+        :param offset: byte of the first character of line in the file.
+        :param line_content: the content of the line.
+        :param line_source: obeject of LineSource.
```
Comment (konefalg) [inline whylog/front/utils.py]: typo: replace "obeject" with "object"

File: .gitignore [hunk 19]
```diff
@@ -58,3 +58,6 @@ target/
 
 # Pycharm
 .idea/
+
+# Whylog
+.whylog_test
```
Comment (kgromadzki) [inline .gitignore]: You should never see this directory. Mea culpa. A forgot write handler which delete this directory if ConfigFactory tests fail. I added it in #101. If you got this directory it means something gone wrong. .gitignore can hide this.

File: whylog/front/whylog_factory.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
+from whylog.config import YamlConfigFactory
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: Front shall not know whether YamlConfigFactory or DbConfigFactory or whatever is used. Loading configuration is a responsibility of Config layer.

File: whylog/teacher/__init__.py [hunk 21]
```diff
@@ -1,4 +1,4 @@
-from whylog.teacher.constraint_links_base import ConstraintLinksBase
+# from whylog.teacher.constraint_links_base import ConstraintLinksBase
```
Comment (ppolewicz) [inline whylog/teacher/__init__.py]: Commented code

File: whylog/teacher/__init__.py [hunk 22]
```diff
@@ -39,7 +39,7 @@ def __init__(self, config, pattern_assistant):
 
         self._parsers = {}
         self._constraint_base = {}
-        self._constraint_links = ConstraintLinksBase()
+        # self._constraint_links = ConstraintLinksBase()
         self.effect_id = None
```
Comment (ppolewicz) [inline whylog/teacher/__init__.py:45]: Commented code

File: whylog/front/__init__.py [hunk 23]
```diff
@@ -1,5 +0,0 @@
-class FrontInput(object):
-    def __init__(self, offset, line_content, resource_location):
-        self.offset = offset
-        self.line_content = line_content
-        self.resource_location = resource_location
```
Comment (kgromadzki) [inline whylog/front/__init__.py:5]: Changed every occurence of resource location to line_source

**Comment by ppolewicz** [comment]
Inappropriate commit comment. Please reword it.

File: whylog/front/whylog_factory.py [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+from whylog.teacher import Teacher
+
+
+def whylog_factory():
+    """
+    This function creates object of YamlConfig and use this object
+    to creates the LogReader object and teachers_generator.
+    """
+
+    class TeacherGenerator():
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py]: `functools.partial`

File: whylog/front/whylog_factory.py [hunk 25]
```diff
@@ -0,0 +1,22 @@
+from whylog.config import SettingsFactorySelector
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+
+
+def whylog_factory():
+    """
+    This function creates object of YamlConfig and use this object
```
Comment (epawlowska) [inline whylog/front/whylog_factory.py]: Shouldn't it be Config instead of YamlConfig?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+        log_reader, teacher_generator = whylog_factory()
+        teacher = teacher_generator()
+
+        # check if returned object classes
+        assert isinstance(log_reader, LogReader)
+        assert isinstance(teacher, Teacher)
+    except Exception as exc:
+        raise exc
+    finally:
+        path = os.getcwd()
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Maybe rename "path" to sth like "current_directory_path"?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+        config_dir = os.path.join(path, '.whylog_test')
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Maybe rename '.whylog_test' to '.temp_config_dir"? I don't see connection between config_dir and ".whylog_test"

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+    try:
+        log_reader, teacher_generator = whylog_factory()
+        teacher = teacher_generator()
+
+        # check if returned object classes
+        assert isinstance(log_reader, LogReader)
+        assert isinstance(teacher, Teacher)
+    except Exception as exc:
+        raise exc
+    finally:
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Please, tell me what do we test in following section. I see only removing some directory. Why do we test it?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+        path = os.getcwd()
+        config_dir = os.path.join(path, '.whylog_test')
+        if os.path.isdir(config_dir):
+            shutil.rmtree(config_dir)
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Why do we want to delete this file? We didn't created it in this test. It shouldn't exist. Or if it exists, it means someone else created it. Should we delete it then?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+    def __call__():
+        pass
+
+
+def tests_whylog_factory():
+    try:
+        log_reader, teacher_generator = whylog_factory()
+        teacher = teacher_generator()
+
+        # check if returned object classes
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Maybe: "check if returned object is instance of proper class" ? I suggest to remove this comment, because we can see below clearly that we are checking classes.

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+        assert isinstance(log_reader, LogReader)
+        assert isinstance(teacher, Teacher)
+    except Exception as exc:
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: What kind of Exception do you expect? Maybe more specific exception catching?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+        raise exc
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Why do you raise exception which is already cought?

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 33]
```diff
@@ -0,0 +1,27 @@
+import os
+import shutil
+from whylog.front.whylog_factory import whylog_factory
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+
+
+class MockClass():
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: I can't see any use of this MockClass. It seems to be dead code. Remove it or add docstring about its purpose.

File: whylog/tests/utils.py [hunk 34]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
...
+        if os.path.isdir(config_dir):
+            shutil.rmtree(config_dir)
+
+    @classmethod
+    def remove_settings_dirs(cls):
+        path = os.getcwd()
+        cls.remove_one(path)
+        for i in range(2):
+            path, _ = os.path.split(path)
+            cls.remove_one(path)
```
Comment (konefalg) [inline whylog/tests/utils.py]: How do you guarantee, that this will not remove user home directory? If I understand this code correctly, if starting path is for example: `~/foo/bar`, you will attempt to remove at first `~/foo/bar` (line 39), then `/foo` (line 42, first time) and finally `~` (line 42 second time)!

Comment (konefalg) [inline whylog/tests/utils.py]: Ouch, I understood. You always append `.whylog_test` to the path and this guarantee that you will never remove user home directory. But how do you ensure that 2 times of doing `cd ..` is enough to remove all temporary files made by tests?

File: whylog/tests/utils.py [hunk 35]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
         rules_path = os.path.join(prefix_path, 'rules.yaml')
         log_type_path = os.path.join(prefix_path, 'log_types.yaml')
         return [parsers_path, rules_path, log_type_path]
+
+
+class TestRemovingSettings(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        #This change was caused by fails os.names() for new directories.
```
Comment (epawlowska) [inline whylog/tests/utils.py:34]: Which change? I suppose this comment is for reviewers because it says about "change". Maybe remove it?

File: whylog/tests/utils.py [hunk 36]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
...
+
+    @classmethod
+    def remove_one(clas, path):
+        config_dir = os.path.join(path, SettingsFactorySelector.WHYLOG_DIR)
+        if os.path.isdir(config_dir):
+            shutil.rmtree(config_dir)
+
+    @classmethod
+    def remove_settings_dirs(cls):
+        path = os.getcwd()
```
Comment (epawlowska) [inline whylog/tests/utils.py]: "path" is too general. Please rename.

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 37]
```diff
@@ -0,0 +1,13 @@
+from whylog.front.whylog_factory import whylog_factory
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+from whylog.tests.utils import TestRemovingSettings
+
+
+class TestWhylogFactory(TestRemovingSettings):
```
Comment (epawlowska) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Why it inherits form TestRemovingSettings?

File: whylog/tests/utils.py (hunk 35 above)
Comment (andrzejgorski) [inline whylog/tests/utils.py:34]: This is moved accepted code.

File: whylog/tests/utils.py (hunk 36 above)
Comment (andrzejgorski) [inline whylog/tests/utils.py]: This is moved accepted code.

File: whylog/tests/utils.py (hunk 34 above)
Comment (andrzejgorski) [inline whylog/tests/utils.py]: This is moved accepted code.

File: whylog/tests/utils.py [hunk 38]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
...
+        cls.remove_one(path)
+        for i in range(2):
```
Comment (epawlowska) [inline whylog/tests/utils.py:55]: Maybe add comment why it is exactly '2'? I suggest to write something about structure of dirs.

File: whylog/tests/utils.py [hunk 39]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
...
+        cls.remove_settings_dirs()
+
+    @classmethod
+    def remove_one(clas, path):
+        config_dir = os.path.join(path, SettingsFactorySelector.WHYLOG_DIR)
+        if os.path.isdir(config_dir):
+            shutil.rmtree(config_dir)
+
+    @classmethod
+    def remove_settings_dirs(cls):
```
Comment (epawlowska) [inline whylog/tests/utils.py]: What does this method? Ir removes 2 latest directory on current directory path? Is it safe? If yes, maybe write comment why is it safe?

File: whylog/tests/utils.py [hunk 40]
```diff
@@ -8,3 +11,33 @@ def get_path_to_config_files(cls, prefix_path):
...
+        SettingsFactorySelector.WHYLOG_DIR = '.whylog_test'
+        cls.remove_settings_dirs()
+
+    @classmethod
+    def tearDownClass(cls):
+        #Removed test config directories if test failed
+        cls.remove_settings_dirs()
+
+    @classmethod
+    def remove_one(clas, path):
```
Comment (epawlowska) [inline whylog/tests/utils.py]: Please, rename it. From name it seems that it removes directory on "path". It looks very dangerous in following `remove_settings_dirs`

File: whylog/tests/utils.py (hunk 35 above)
Comment (andrzejgorski) [inline whylog/tests/utils.py:34]: This is about change name from .whylog to .whylog_test, which is on line 22.

File: whylog/tests/tests_front/tests_whylog_factory.py [hunk 41]
```diff
@@ -0,0 +1,13 @@
...
+    def tests_whylog_factory(self):
+        log_reader, teacher_generator = whylog_factory()
+        teacher = teacher_generator()
+
+        assert isinstance(log_reader, LogReader)
```
Comment (ppolewicz) [inline whylog/tests/tests_front/tests_whylog_factory.py]: Test behavior, not identity. One day someone will make the factory return ThreadedLogReader by default and then this test will fail without reason - ThreadedLogReader will keep the same interface, after all. By creating tests that rely on identity of objects, you make them fragile. Soon when modifying code you will find that by yourself.

File: whylog/front/whylog_factory.py [hunk 42]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def whylog_factory():
+    """
+    This function returns LogReader object and teachers_generator.
+    """
+    settings = SettingsFactorySelector.get_settings()
+    pattern_assistant = settings['assistant']()
+    config = settings['config']
+    return LogReader(config), partial(Teacher, config, pattern_assistant)
```
Comment (ppolewicz) [inline whylog/front/whylog_factory.py:15]: nice!

File: whylog/__init__.py [hunk 43]
```diff
@@ -0,0 +1,14 @@
+from whylog.config import LineSource
+from whylog.config.filename_matchers import RegexFilenameMatcher
+from whylog.config.log_type import LogType
+from whylog.front.utils import FrontInput
+from whylog.front.whylog_factory import whylog_factory
+from whylog.log_reader import LogReader
+from whylog.teacher import Teacher
+assert FrontInput
```
Comment (ppolewicz) [inline whylog/__init__.py:9]: nice! optimized python will cut it out in precompilation

File: .gitignore [hunk 44]
```diff
@@ -58,3 +58,6 @@ target/
 
 # Pycharm
 .idea/
+
+# Whylog
+*.whylog
```
Comment (ppolewicz) [inline .gitignore]: what is that? Why is this here? I don't think it should be here...

File: whylog/tests/utils.py (hunk 36 above)
Comment (ppolewicz) [inline whylog/tests/utils.py]: In that case find out who reviewed it and point it out to him. But fix the variable name as Ewa said. Just because some bad code is present in the repository does not mean that we should not fix it when we spot it.

File: whylog/__init__.py (hunk 43 above)
Comment (andrzejgorski) [inline whylog/__init__.py:9]: But pyflakes8 will not make noise of these lines.
Check: 
http://stackoverflow.com/questions/5033727/how-do-i-get-pyflakes-to-ignore-a-statement
http://www.laurivan.com/make-pyflakespylint-ignore-unused-imports/
I don't know the better solution. If you have better solution give it here.

File: whylog/tests/tests_front/tests_whylog_factory.py (hunk 41 above)
Comment (andrzejgorski) [inline whylog/tests/tests_front/tests_whylog_factory.py]: This test tests only whylog_factory which creates Only LogReader. We don't have implemented ThreadedLogReader yet and we don't have implemented LogReaderSelector or something like this. I only use the interface of LogReader that is available. If something is wrong with it give comments not to me but to @konefalg .

File: whylog/__init__.py (hunk 43 above)
Comment (ppolewicz) [inline whylog/__init__.py:9]: As I said, it is very nice solution of this problem

**Comment by ppolewicz** [comment]
please merge to resolve conflict and lets merge it.
